### PR TITLE
Setting all configuration attributes by default is dangerous, so requ…

### PR DESCRIPTION
**Pull request by splatteredbits**
…ire a switch to do it.

### CODE REVIEW

File: Carbon.IIS/Functions/Install-CIisAppPool.ps1 [hunk 1]
```diff
@@ -8,12 +8,9 @@ function Install-CIisAppPool
...
     The `Install-CIisAppPool` function creates or updates an IIS application pool. Pass the name of the application pool
     to the `Name` parameter. If that application pool doesn't exist, it is created. If it does exist, its configuration
-    is updated to match the values of the arguments passed.
-
-    If you pass just a name, the function creates a 64-bit application pool that runs an integrated pipeline
-    using .NET 4.0 managed runtime. To use a 32-bit app pool, set the `Enable32BitAppOnWin64` parameter to `$true`. To use a
-    classic pipeline, set the `ManagedPipelineMode` parameter to `Classic`. To use a different version of .NET, use the
-    `ManagedRuntimeVersion` parameter.
+    is updated to match the values of the arguments passed. If you don't pass an argument, that argument's setting is
+    deleted and reset to its default value. You always get an application pool with the exact same configuration, even
```
Comment (pnstickne) [inline Carbon.IIS/Functions/Install-CIisAppPool.ps1:12]: "deleted and reset to its default value" appears to disagree with `-Reset` notes.

File: Carbon.IIS/Functions/Set-CIisAppPoolPeriodicRestart.ps1 [hunk 2]
```diff
@@ -15,23 +15,32 @@ function Set-CIisAppPoolPeriodicRestart
...
-
     .EXAMPLE
     Set-CIisAppPoolPeriodicRestart -AppPoolName 'Snafu' -Memory 1000000 -PrivateMemory 2000000 -Requests 3000000 -Time '23:00:00'
 
     Demonstrates how to configure all an IIS applicaton pool's periodic restart settings. In this example, `memory` will
     be set to `1000000`, `privateMemory` will be set to `2000000`, `requests` will be sent to `3000000`, and `time` will
     be sent to `23:00:00'.
+
+    .EXAMPLE
+    Set-CIisAppPoolPeriodicRestart -AppPoolName 'Fubar'-Memory 1000000 -PrivateMemory 2000000 -Reset
```
Comment (pnstickne) [inline Carbon.IIS/Functions/Set-CIisAppPoolPeriodicRestart.ps1]: Missing space after `'Fubar'`.

Also perhaps different name 🍤

File: Carbon.IIS/Functions/Set-CIisAppPoolPeriodicRestart.ps1 [hunk 3]
```diff
@@ -102,13 +117,14 @@ function Set-CIisAppPoolPeriodicRestart
...
             }
             if( -not $clearedPrefix )
             {
                 $prefixMsg = ' ' * $prefixMsg.Length
                 $clearedPrefix = $true
             }
-    }
+        }
 
         $currentSchedule.Clear()
```
Comment (pnstickne) [inline Carbon.IIS/Functions/Set-CIisAppPoolPeriodicRestart.ps1]: `if( $scheduleChanged )` around all this processing as well, to not even modify-in-memory if not changed?

File: Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1 [hunk 4]
```diff
@@ -74,17 +74,31 @@ function Set-CIisConfigurationAttribute
...
 
+        # Properties to skip and not change. These are usually private settings that we shouldn't be mucking with or
+        # settings that capture current state, etc.
+        [Parameter(ParameterSetName='AllByConfigElement')]
+        [Parameter(ParameterSetName='AllByConfigPath')]
+        [String[]] $Exclude = @(),
+
+        # Resets *all* the target configuration element's settings to their default values, *except* each setting
+        # passed via the `Attribute` parameter hashtable. The default behavior is to only modify each setting passed in
+        # The `Attribute` parameter hashtable.
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1]: Typo: `The` should not be capitalized.

File: Carbon.IIS/Functions/Set-CIisAppPoolProcessModel.ps1 [hunk 5]
```diff
@@ -111,7 +119,14 @@ function Set-CIisAppPoolProcessModel
         return
     }
 
+    $targetMsg = 'IIS application pool defaults process model'
```
Comment (pnstickne) [inline Carbon.IIS/Functions/Set-CIisAppPoolProcessModel.ps1:122]: Perhaps Invoke-SetConfigurationAttribute could take in, eg. `-Target @(message, scope)` , where scope is the identifier.

File: Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1 [hunk 6]
```diff
@@ -74,17 +74,31 @@ function Set-CIisConfigurationAttribute
...
         [Parameter(ParameterSetName='AllByConfigPath')]
         [String] $Target,
 
+        # Properties to skip and not change. These are usually private settings that we shouldn't be mucking with or
+        # settings that capture current state, etc.
+        [Parameter(ParameterSetName='AllByConfigElement')]
+        [Parameter(ParameterSetName='AllByConfigPath')]
+        [String[]] $Exclude = @(),
+
+        # Resets *all* the target configuration element's settings to their default values, *except* each setting
```
Comment (pnstickne) [inline Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1]: Perhaps include interaction with `-Exclude`.

I might expect exclude to be applied first, including to `-Reset`.

**Review by decoyjoe — APPROVED** [review]

File: whiskey.yml [hunk 7]
```diff
@@ -3,7 +3,10 @@ Build:
 - Version:
     Path: Carbon.IIS\Carbon.IIS.psd1
     Prerelease:
-    - "*": rc1
+    - "prerelease/*": rc1
+    - "feature/*": alpha1
+    - "bugfix/*": beta1
```
Comment (pnstickne) [inline whiskey.yml:8]: bugfix feels odd as beta given others. Can't quite nail it down.

File: whiskey.yml [hunk 8]
```diff
@@ -3,7 +3,10 @@ Build:
 - Version:
     Path: Carbon.IIS\Carbon.IIS.psd1
```
Comment (pnstickne) [inline whiskey.yml:4]: Version not bumped?

**Review by pnstickne — APPROVED** [review]
Approved with notes.

Version not changed?

**Comment by splatteredbits** [comment]
> Approved with notes.
> 
> Version not changed?

I've got a bunch more changes coming. These will be released as an RC until I get all the new functionality needed in the test environment.

File: whiskey.yml (hunk 7 above)
Comment (splatteredbits) [reply whiskey.yml:8]: Bugfixes usually happen off already-released code, so I thought beta was the best modifier.

File: Carbon.IIS/Functions/Install-CIisAppPool.ps1 (hunk 1 above)
Comment (splatteredbits) [reply Carbon.IIS/Functions/Install-CIisAppPool.ps1:12]: Changed to "deleted, which resets it to its default value". I like that wording better, so updated all the `-Reset` docs to match.

File: Carbon.IIS/Functions/Set-CIisAppPoolPeriodicRestart.ps1 (hunk 3 above)
Comment (splatteredbits) [reply Carbon.IIS/Functions/Set-CIisAppPoolPeriodicRestart.ps1]: Good catch.

File: Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1 (hunk 6 above)
Comment (splatteredbits) [reply Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1]: `Reset` respects `Exclude`. The `Set-AttributeValue` function is where the Exclude parameter is handled and here's where we use `Reset` and `Set-AttributeValue`:

```powershell
        $ConfigurationElement.Attributes |
            Where-Object {
                if( $Reset )
                {
                    return $true
                }

                return $Attribute.ContainsKey($_.Name)
            } |
            Sort-Object -Property 'Name' |
            Set-AttributeValue
```

File: whiskey.yml (hunk 8 above)
Comment (splatteredbits) [reply whiskey.yml:4]: We haven't released 1.0.0 of this module yet. It was extracted from Carbon.

**Pull request merged**